### ISSUE TITLE
Classes defined with an abstract base class's `register` method fail type checks

### ISSUE THREAD

**Issue by bavalpey** [microsoft/pyright#8139]
**Describe the bug**
Abstract base classes provide a `register` method that, among other things, allows classes and their instances to pass `issubclass()` and `isinstance()` checks, respectively, against the ABC. Pyright does does not detect this and reports an incompatible type.

**Code or Screenshots**

```python
import abc

class Foo(abc.ABC):
    @abc.abstractmethod
    def bar(self):
        pass

@Foo.register  # alternatively, Foo.register(Qux) after its definition
class Qux:
    def bar(self):
        pass
def quux() -> Foo:
    return Qux()
```

![image](https://github.com/microsoft/pyright/assets/36710691/652ea45a-8626-4cf5-84b6-2048fd04fdab)

**VS Code extension or command-line**
Pyright cli, version 1.1.367

**Comment by erictraut** [comment]
ABC registration is a dynamic operation that cannot be supported by static type checkers. If you want your code to work with static type checking, you'll need to modify it to use direct inheritance of the ABC. Protocol classes may also be of use.

**Comment by bavalpey** [comment]
> ABC registration is a dynamic operation that cannot be supported by static type checkers. If you want your code to work with static type checking, you'll need to modify it to use direct inheritance of the ABC. Protocol classes may also be of use.

The use case for this is an Enum which cannot derive from ABC. Do you know of an alternative to get it to satisfy pyright? (I was worried this was dynamic anyway).

**Comment by erictraut** [comment]
I don't know enough about your use case, but as I mentioned above, you may be able to use a Protocol class in place of an ABC.

**Comment by fergusq** [comment]
@erictraut Could you reconsider this feature? While ABC registration is dynamic, so are most things in the Python type system. I don't see how ABC registration is so different from e.g. dataclass decorators, that the type checker can handle just fine.

The basic use for this feature would be creating a virtual superclass for Enums or NamedTuples that cannot inherit classes. See for example:

```python
class Weight(ABC):
    @abstractmethod
    def as_int(self) -> int:
        ...

@Weight.register
class IntWeight(NamedTuple):
    data: int
    def as_int(self):
        return self.data

@Weight.register
class FloatWeight(NamedTuple):
    data: float
    def as_int(self):
        return int(self.data)
```

A protocol does not support proper runtime type-checking: with `@runtime_checkable`, it might recognize some unrelated classes that just happen to have the same method name `as_int`, breaking `isinstance` calls intended to check exactly for these two classes and nothing else (in the case of the example above, imagine that there were also ABCs `Length`, `Duration` etc. with the same `as_int` method and implementations for them).

Currently, the only way to implement this is to not use Enum or NamedTuple. Frozen dataclasses work in the situation above. But there are multiple good reasons to use Enums and NamedTuples and it feels hacky to use dataclasses.

Since dataclasses are a runtime feature too, I don't really see how checking for the ABC decorator is much different from checking the dataclass decorator.

**Comment by bavalpey** [comment]
> @erictraut Could you reconsider this feature? While ABC registration is dynamic, so are most things in the Python type system. I don't see how ABC registration is so different from e.g. dataclass decorators, that the type checker can handle just fine.

In my experience, there's little use petitioning the maintainers of this repo once they've said no to something. It's best to put your requests and effort towards more community-friendly projects. BasedPyright, and now Astral's Ty are much better (at least for now) at engaging and working with their community userbase.
